clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -762,7 +764,7 @@ synchronized void maybeResolveSequences() {
                         // For the transactional producer, we bump the epoch if possible, otherwise we transition to a fatal error
                         String unackedMessagesErr = "The client hasn't received acknowledgment for some previously " +
                                 "sent messages and can no longer retry them. ";
-                        if (canBumpEpoch()) {
+                        if (!isTransactionV2Enabled && canBumpEpoch()) {
```

Comment:
I don't think this logic is correct (in all places).  As currently written with v2 we'll fall through to transition to fatal error.  We still want to transition to abortable error just without epoch bumping.   So the logic should look like this (in all places).

```
if (canBumpEpoch()) {
 if (!isTransactionV2Enabled) epochBumpRequired = true;
   // ... whatever logic was there unchanged ...
 } else {
   // ... whatever logic was there unchanged ...
 }
```

but instead of writing this in all places refactor into method `maybeEpochBumpRequired`  In that method we can have a comment that explains why we don't need an explicit epoch bump from the client: with v2 the epoch will get bumped on abort automatically.

Reply:
I originally suggested moving the transactionsV2Enabled into the canBumpEpoch check, but I see now that this wouldn't work for the suggestion. I wonder if we should rename that method to "canAbort" or something that clearly indicates what it will be signaling now.

Reply:
Agree on figuring out a better name, we now have many levels for supporting epoch bumps in the system!  My understanding is that canBumpEpoch checks whether the broker supports bumping epoch via calling InitProducerId.  And transactionsV2Enabled means that broker supports bumping epoch on each transaction (which is the next level of epoch bumping that eliminates the need of manual epoch bumping).

canAbort is probably fine (without epoch bump support in some form we really cannot abort transactions without risk of getting zombies).  For idempotent producer, though, it doesn't make sense, but maybe we should go through the code and don't use this check for code paths for idempotent producer.  Basically

- rename canBumpEpoch to canAbort and remove the logic for idempotent producer from that method
- modify all places where canBumpEpoch is called for idempotent producer to not call (because it's always true for idempotent)

This makes me think (and I know -- it was me who suggested to change the epoch bumping change in this PR, but I didn't realize it becomes tricky) maybe we should remove manual epoch bumping in a separate PR to keep this PR small and unblock other work.

Reply:
Seems reasonable to me. Do we foresee any issues with not including this until later? Mostly wondering if we will have correctness issues and if we should delay marking TV2 as ready before making these changes.

Reply:
I cannot think of a correctness issue resulting from redundant epoch bump, but out of caution I'd suggest that we get the change in before making TV2 ready, I think it's just going to be easier to review as a separate PR (and merging this PR would also unblock the PR where we remove explicit add partitions call).

So call this PR [2/3] and then have another PR [3/3] to remove manual epoch bumping logic.

Reply:
> As currently written with v2 we'll fall through to transition to fatal error.

Reply:
Maybe we should also rename epochBumpReuired to manualEpochBumpRequired or clientInitiatedEpochBumpRequired, and instead of using transactionV2Enabled as a flag everywhere, we use either manual or automatic epoch bumps as conditions. In the comments we can specify that this is only possible when V2 is enabled etc. And we should probably be able to move the logic of "coordinatorSupportsBumpingEpoch" also into one of these categories or merge logic for that as well.

Reply:
@artemlivshits We can do it in a separate PR but there are still gonna be test failures :( Since we're pretty close to resolving the issues I'm wondering if we should just do everything in this PR and make sure all the tests pass.

Reply:
Artem was saying he thought there wouldn't be any correctness issues, so I'm wondering if there will be test failures. Did you see that this change fixed them?

Reply:
We saw failures in the integration test due to incorrectly setting the lastEpoch to the current epoch in the prepareComplete method, this caused an epoch fenced error when initProducerId was tried with the original epoch after an abort transaction already updated the epoch to original + 1.